codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Headers.java

```diff
@@ -354,4 +285,34 @@ public ByteString path() {
     public ByteString status() {
         return get(PseudoHeaderName.STATUS.value());
     }
+
+    private static class Http2HeaderNameComparator implements Comparator<ByteString>, Serializable {
+
+        private static final Http2HeaderNameComparator INSTANCE = new Http2HeaderNameComparator();
+        private static final long serialVersionUID = 1109871697664666478L;
+
+        @Override
+        public int compare(ByteString one, ByteString two) {
+            // Reserved header names come first.
+            final boolean isPseudoHeader1 = one.byteAt(0) == ':';
+            final boolean isPseudoHeader2 = two.byteAt(0) == ':';
+            if (isPseudoHeader1 && !isPseudoHeader2) {
+                return -1;
+            }
+            if (!isPseudoHeader1 && isPseudoHeader2) {
+                return 1;
+            }
+            final int delta = one.hashCode() - two.hashCode();
+            if (delta == 0) {
+                // If the hash code matches it's very likely for the two strings to be equal
+                // and thus we optimistically compare them with the much faster equals method.
+                if (one.equals(two)) {
```

Comment:
> If we implemented the compare method on ByteString, we could use it here.

@nmittler this call to `equals()` is actually a call to my unsafe version of equals. So this mostly to improve performance. That's mostly why the HTTP/2 is faster than HTTP for lookup / insertion.